### PR TITLE
Feature engine

### PR DESCRIPTION
**Pull request by rk4514**
This is the mostly completed game engine. Only thing left to do is implement trading and communication with sockets. I've included some functions that interface with the console for the sake of testing but those will be removed later.

### CODE REVIEW

File: batan-server/engine/board.ts [hunk 1]
```diff
@@ -158,11 +160,34 @@ export class Board {
...
+        this.vertexList[42].harbor = harborType.sheep;
+        this.vertexList[46].harbor = harborType.sheep;
+        this.vertexList[49].harbor = harborType.threeForOne;
+        this.vertexList[52].harbor = harborType.threeForOne;
+        this.vertexList[51].harbor = harborType.threeForOne;
+        this.vertexList[47].harbor = harborType.threeForOne;
+        this.vertexList[38].harbor = harborType.brick;
+        this.vertexList[33].harbor = harborType.brick;
+        this.vertexList[16].harbor = harborType.wood;
+        this.vertexList[11].harbor = harborType.wood;
```
Comment (jackadrianglass) [inline batan-server/engine/board.ts:184]: Could this be put into a map and iterated through? Like {idx: type}

File: batan-server/engine/board.ts [hunk 2]
```diff
@@ -242,97 +241,97 @@ export class Board {
...
 
     /**
      * Set Vertices owned by each Hex.
      */
-    private setHexVertices(){
-        for(let i = 0; i < 3; i++){
-            this.hexList[i].vertices = [0+1, 3+i, 4+i, 7+i, 8+i, 12+i];
+    private setHexVertices() {
+        for (let i = 0, j = 0; i < 3; i++, j++) {
+            this.hexList[i].vertices = [0 + j, 3 + j, 4 + j, 7 + j, 8 + j, 12 + j];
```
Comment (jackadrianglass) [inline batan-server/engine/board.ts]: couldn't you use i instead of j? Their the same value the entire time

**Review by jackadrianglass — APPROVED** [review]
Looks good to me. I had a couple of nits

File: batan-server/engine/game.ts [hunk 3]
```diff
@@ -99,93 +194,494 @@ class Game {
...
+    //const diceRoll = roll;
+
+    console.log(`${player[user]} rolled a ${diceRoll}!`);
 
     if (diceRoll === 7) {
-      //move robber logic and take resources
+      this.moveRobberAndSteal(user);
+
+      //discard if more than 7
+      this.players.forEach((player) => {
```
Comment (lhossackucalgary) [inline batan-server/engine/game.ts:474]: Robber should be moved _after_ players discard half their cards if they have over 7

File: batan-server/engine/game.ts [hunk 4]
```diff
@@ -99,93 +194,494 @@ class Game {
...
+        let totalResources = player.resources.brick + player.resources.ore + player.resources.sheep + player.resources.wheat + player.resources.wood;
+        if (totalResources >= 7) {
+          //TODO emit to user... this will be handled with sockets later ... need to ensure correct number discarded
```
Comment (lhossackucalgary) [inline batan-server/engine/game.ts]: I think we'll probably need to split this into a few functions or include an additional state machine for context of being part way through this operation during the back & forth between client/server

**Review by lhossackucalgary — APPROVED** [review]
lgtm for this merge. 
For the future, I have some concerns about some of the operations like when a 7 is rolled (& robber is moved) that require user input part way through the operation. I think it can be resolved by adding some additional state that makes other moves invalid until everyone with over 7 cards has tossed half their cards for example, but it definitely throws a bit of a wrench into the strictly turn-based flow (& trading will as well). I don't know if we will even get to implementing trading on the front end with the 10 days we have remaining though

File: batan-server/engine/game.ts (hunk 4 above)
Comment (rk4514) [reply batan-server/engine/game.ts]: @lhossackucalgary for sake of time, we can just make it take the appropriate number of cards at random if that is easier. Should not affect game play all that much.

File: batan-server/engine/board.ts (hunk 1 above)
Comment (rk4514) [reply batan-server/engine/board.ts:184]: Yes but for sake of time i'm going to leave this as is.

File: batan-server/engine/game.ts (hunk 3 above)
Comment (rk4514) [reply batan-server/engine/game.ts:474]: Done, nice catch!

**Comment by rk4514** [comment]
@jackadrianglass @lhossackucalgary Thanks for the feedback! Comments addressed.

**Pull request merged**